### PR TITLE
New major version

### PR DESCRIPTION
**Pull request by woksin**
I did lots of changes, but most of them are trivial and similar
Added:
* Exception and NotImplemented to 'rudiments'
* Added Extension to 'rudiments' which is a decorator that can be used to augment classes with extension methods.

Major Changes:
* 'Cause' static fromReason method instead of withArguments and noArguments methods on 'Reason'. For one I think it makes more sense and the other reason is that it hid a circular dependency which caused problems when I fixed the import statements in the specs and code. 
* 'Reason' _id is now of type 'Guid' - more explicit

Fixes:
* Standardised all import statements. Code in a module refer to each other through the modules that expose them. Specs always refer to the files directly inside the same package.

Issues:
I had to do something weird to get the extension methods for 'PropertyRuleSetBuilder' to work. It builds and the tests passes, but wallaby has trouble with running it. When building it normally it works fine it is just the transpiled version for wallaby that does not work for this. I expect that it is because of something weird with that outputted code. It should be fixed if we decide to transpile code for wallaby using a more common target.

### CODE REVIEW

**Comment by woksin** [comment]
https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/7/commits/f56edfa543093715b3c8456906d5fe057391af03 seems to fix the problem I had with wallaby. Now it works like a charm for me.

Can you maybe try this code out, @einari and confirm that it does not break anything you have?

File: Source/rudiments/Extension.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Dolittle. All rights reserved.
```
Comment (einari) [inline Source/rudiments/Extension.ts]: What is the benefit of this over:

```typescript
Something.prototype.doSomething = function(thisArg: Something): any {
};
```

With this approach there is more you need to do to do an extension with the construct of a class, static method and a decorator.

File: Source/reactive/package.json [hunk 2]
```diff
@@ -1,38 +1,38 @@
 {
-    "name": "@dolittle/reactive",
```
Comment (einari) [inline Source/reactive/package.json:2]: I see that we're missing an editor config file in this repo - we normally have 4 in indent size. Great if you can add an `.editorconfig` at the root of the project that looks like below and reformat the package.json files.

```ini
# EditorConfig helps create consistency for formatting of all 
# the code in Dolittle
#
# http://EditorConfig.org

root = true

[*]
end_of_line = lf
indent_style = space
indent_size = 4
charset = utf-8

[*.yml]
indent_size = 2
```

File: Source/rules/for_ObjectRuleSetContainerBuilder/when_building_with_one_property_added.ts [hunk 3]
```diff
@@ -2,12 +2,11 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 import { an_object_with_nested_properties } from './given/an_object_with_nested_properties';
-import { PropertyRuleSetBuilder } from '../index';
 
 describe('when building with one property added', () => {
     const given: an_object_with_nested_properties = new an_object_with_nested_properties();
 
-    given.builder.rulesFor(_ => _.secondLevel);
+    given.builder.rulesFor((_: any) => _.secondLevel);
```
Comment (einari) [inline Source/rules/for_ObjectRuleSetContainerBuilder/when_building_with_one_property_added.ts]: The type here is something the TypeScript compiler is able to derive, it becomes noisy having to explicitly add `(_: any)`.

**Review by einari — CHANGES_REQUESTED** [review]
Summary from review:

Formatting, we use 4 in indent-size, so the `package.json` files can be reverted and suggest you introduce  a `.editorconfig` file at the root: 

```ini
# EditorConfig helps create consistency for formatting of all 
# the code in Dolittle
#
# http://EditorConfig.org

root = true

[*]
end_of_line = lf
indent_style = space
indent_size = 4
charset = utf-8

[*.yml]
indent_size = 2
```

I see the introduction of `@extension` decorator, but feel that the construct for this in JavaScript should suffice and this makes it more complicated.

A little bit of a summary from a discussion done on Teams and not here:

The proposed API changes to Cause / Reason. I think this should remain as is in the sentiment; meaning that you create a cause from a reason and Reason is responsible for doing so. The relationship between Cause and Reason is important as a Cause is an instance of a Reason and should point back to the reason it came from. How the API for creating a cause instance could probably be a lot better the `noArguments()` and `withArguments()` is not inviting to clear understanding as it was. 

The change of type for the unique identifier for Reason adds a breaking change to the API as well. 
Suggest adding a ReasonId type which can be either a `Guid` or a `string`- suggest looking [here](https://github.com/dolittle-runtime/JavaScript.SDK/blob/master/Source/artifacts/ArtifactId.ts) which defines a type which can be both. And then use the `.as()` method on `Guid` internally to make it `Guid`. An example can be found [here](https://github.com/dolittle-runtime/JavaScript.SDK/blob/54f6bdf3d5910eda5635ac17f69fa7a01c69fb03/Source/artifacts/Artifacts.ts#L81)

File: Source/reactive/package.json (hunk 2 above)
Comment (woksin) [reply Source/reactive/package.json:2]: Resolved by https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/7/commits/5e7558555ae66ffa181659e003cccb2c67a58c30

File: Source/rudiments/Extension.ts (hunk 1 above)
Comment (woksin) [reply Source/rudiments/Extension.ts]: Yeah I agree, it was just me exploring a bit. I found another solution to the underlying problem so this can go

Comment (woksin) [reply Source/rudiments/Extension.ts]: resolved by https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/7/commits/8cd49a400fd2471f22474a1c62a95f0811670eed

File: Source/rules/for_ObjectRuleSetContainerBuilder/when_building_with_one_property_added.ts (hunk 3 above)
Comment (woksin) [reply Source/rules/for_ObjectRuleSetContainerBuilder/when_building_with_one_property_added.ts]: Woops, that was a left over thing...
https://github.com/dolittle-fundamentals/JavaScript.Fundamentals/pull/7/commits/1c7d4577b08b15fba66b80ab93e1cf5234f1fd4e resolves it

**Comment by woksin** [comment]
@einari Check that my changes resolved your issues

**Review by einari — APPROVED** [review]

**Pull request merged**